spec/tapioca/compilers/dsl/active_record_associations_spec.rb

```diff
@@ -0,0 +1,317 @@
+  before(:each) do
+    require "tapioca/compilers/dsl/active_record_associations"
+  end
+
+  subject do
+    Tapioca::Compilers::Dsl::ActiveRecordAssociations.new
+  end
+
+  describe("#initialize") do
+    def constants_from(content)
+      with_content(content) do
+        subject.processable_constants.map(&:to_s).sort
+      end
+    end
+
+    it("gathers no constants if there are no ActiveRecord subclasses") do
+      assert_empty(subject.processable_constants)
+    end
+
+    it("gathers only ActiveRecord subclasses") do
+      content = <<~RUBY
+        class Post < ActiveRecord::Base
+        end
+
+        class Current
+        end
+      RUBY
+
+      assert_equal(constants_from(content), ["Post"])
+    end
+
+    it("rejects abstract ActiveRecord subclasses") do
+      content = <<~RUBY
+        class Post < ActiveRecord::Base
+        end
+
+        class Current < ActiveRecord::Base
+          self.abstract_class = true
+        end
+      RUBY
+
+      assert_equal(constants_from(content), ["Post"])
+    end
+  end
+
+  describe("#decorate") do
+    before(:each) do
+      ActiveRecord::Base.establish_connection(
+        adapter: 'sqlite3',
+        database: ':memory:'
+      )
+    end
+
+    def rbi_for(content)
+      with_content(content) do
+        parlour = Parlour::RbiGenerator.new(sort_namespaces: true)
+        subject.decorate(parlour.root, Post)
+        parlour.rbi
+      end
+    end
+
+    it("generates empty RBI file if there are no associations") do
+      content = <<~RUBY
+        class Post < ActiveRecord::Base
+        end
+      RUBY
+
+      expected = <<~RUBY
+        # typed: strong
+
+      RUBY
+
+      assert_equal(rbi_for(content), expected)
+    end
+
+    it("generates RBI file for belongs_to single association") do
+      content = <<~RUBY
+        class Post < ActiveRecord::Base
+          belongs_to :category
+        end
+      RUBY
+
+      expected = <<~RUBY
+        # typed: strong
+        class Post
+          include Post::GeneratedAssociationMethods
+        end
+
+        module Post::GeneratedAssociationMethods
+          sig { params(args: T.untyped, blk: T.untyped).returns(T.nilable(T.untyped)) }
+          def build_category(*args, &blk); end
+
+          sig { returns(T.nilable(T.untyped)) }
+          def category; end
+
+          sig { params(value: T.nilable(T.untyped)).void }
+          def category=(value); end
+
+          sig { params(args: T.untyped, blk: T.untyped).returns(T.nilable(T.untyped)) }
+          def create_category(*args, &blk); end
+
+          sig { params(args: T.untyped, blk: T.untyped).returns(T.nilable(T.untyped)) }
+          def create_category!(*args, &blk); end
+
+          sig { returns(T.nilable(T.untyped)) }
+          def reload_category; end
+        end
+      RUBY
+
+      assert_equal(rbi_for(content), expected)
+    end
+
+    it("generates RBI file for polymorphic belongs_to single association") do
```

Comment:
Can we also have a test that would test the:
```ruby
if reflection.through_reflection?
  polymorphic_association?(reflection.source_reflection)
```
part as well?

Reply:
Defined a table within this test which now goes through that path https://github.com/Shopify/tapioca/pull/70/files#diff-08e12da6ea17b18ee3475e651faf0189R231